[Flaking Test] integration-master (scheduler_perf tests)

### Which jobs are flaking?

master-blocking:
- integration-master

### Which tests are flaking?

`k8s.io/kubernetes/test/integration/scheduler_perf.scheduler_perf`

### Since when has it been flaking?

The only run that failed on master-blocking was: [5/17/2024, 9:22:18 AM](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-integration-master/1791315518829891584)

But [Triage](https://storage.googleapis.com/k8s-triage/index.html?test=k8s.io%2Fkubernetes%2Ftest%2Fintegration%2Fscheduler_perf.scheduler_perf) shows this test to be failing across different jobs with an extremely high rate. 

### Testgrid link

https://testgrid.k8s.io/sig-release-master-blocking#integration-master

### Reason for failure (if possible)

The complete build log can be found here: https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-integration-master/1795650509529419776/build-log.txt

There are numerous errors, highlighting some of them:

```
E0529 03:48:46.793158  117977 controller.go:159] "Unhandled Error" err="unable to sync kubernetes service: Endpoints \"kubernetes\" is invalid: subsets[0].addresses[0].ip: Invalid value: \"127.0.0.1\": may not be in the loopback range (127.0.0.0/8, ::1/128)" logger="UnhandledError"
...
    schedule_one.go:1052: E0529 03:48:37.463675] Error scheduling pod; retrying err="waiting for resource driver" pod="test/test-drac5bhg"
...
E0529 03:09:54.849485   25438 reflector.go:158] "Unhandled Error" err="k8s.io/client-go/metadata/metadatainformer/informer.go:138: Failed to watch *v1.PartialObjectMetadata: failed to list *v1.PartialObjectMetadata: the server could not find the requested resource" logger="UnhandledError"
...
E0529 03:08:16.849871   25438 replica_set.go:560] "Unhandled Error" err="sync \"namespace-1716952096-8804/cassandra\" failed with replicationcontrollers \"cassandra\" not found" logger="UnhandledError"
...
E0529 03:07:37.755211   25438 replica_set.go:236] "Unhandled Error" err="ReplicaSet has no controller: &ReplicaSet{ObjectMeta:{frontend  namespace-1716952054-23531  228ae60e-495d-445c-955c-2b4280009709 3244 2 2024-05-29 03:07:36 +0000 UTC <nil> <nil> map[app:guestbook tier:frontend] map[] [] [] []},Spec:ReplicaSetSpec{Replicas:*2,Selector:&v1.LabelSelector{MatchLabels:map[string]string{app: guestbook,tier: frontend,},MatchExpressions:[]LabelSelectorRequirement{},},Template:{{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[app:guestbook tier:frontend] map[] [] [] []} {[] [] [{php-redis gcr.io/google_samples/gb-frontend:v3 [] []  [{ 0 80 TCP }] [] [{GET_HOSTS_FROM dns nil}] {map[] map[cpu:{{100 -3} {<nil>} 100m DecimalSI} memory:{{104857600 0} {<nil>} 100Mi BinarySI}] []} [] <nil> [] [] nil nil nil nil /dev/termination-log File IfNotPresent nil false false false}] [] Always 0xc00442e218 <nil> ClusterFirst map[]   <nil>  false false false <nil> PodSecurityContext{SELinuxOptions:nil,RunAsUser:nil,RunAsNonRoot:nil,SupplementalGroups:[],FSGroup:nil,RunAsGroup:nil,Sysctls:[]Sysctl{},WindowsOptions:nil,FSGroupChangePolicy:nil,SeccompProfile:nil,AppArmorProfile:nil,} []   nil default-scheduler [] []  <nil> nil [] <nil> <nil> <nil> map[] [] <nil> nil <nil> [] []}},MinReadySeconds:0,},Status:ReplicaSetStatus{Replicas:3,FullyLabeledReplicas:3,ObservedGeneration:1,ReadyReplicas:0,AvailableReplicas:0,Conditions:[]ReplicaSetCondition{},},}" logger="UnhandledError"
```

### Anything else we need to know?

https://testgrid.k8s.io/sig-scalability-benchmarks#scheduler-perf is 🔴 

### Relevant SIG(s)

/sig scalability
@kubernetes/release-team-release-signal

I also checked two failures in ci-benchmark-scheduler-perf-master side, and all failures in those two were from DRA test cases.

```
=== Failed
=== FAIL: test/integration/scheduler_perf BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplate/2000pods_100nodes (0.00s)
--- FAIL: BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplate/2000pods_100nodes
BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplate/2000pods_200nodes
    scheduler_perf.go:742: disabled by label filter "performance"

=== FAIL: test/integration/scheduler_perf BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplate (0.00s)
--- FAIL: BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplate
BenchmarkPerfScheduling/SchedulingWithMultipleResourceClaims
BenchmarkPerfScheduling/SchedulingWithMultipleResourceClaims/fast
    scheduler_perf.go:742: disabled by label filter "performance"

=== FAIL: test/integration/scheduler_perf BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplateStructured/2000pods_100nodes (0.00s)
--- FAIL: BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplateStructured/2000pods_100nodes
BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplateStructured/2000pods_200nodes
    scheduler_perf.go:742: disabled by label filter "performance"

=== FAIL: test/integration/scheduler_perf BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplateStructured (0.00s)
--- FAIL: BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplateStructured
    scheduler_perf.go:681: full log output:

=== FAIL: test/integration/scheduler_perf BenchmarkPerfScheduling (0.00s)
--- FAIL: BenchmarkPerfScheduling
```

https://storage.googleapis.com/kubernetes-jenkins/logs/ci-benchmark-scheduler-perf-master/1797013157097508864/build-log.txt

```
=== FAIL: test/integration/scheduler_perf BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplateStructured/2000pods_100nodes (0.00s)
--- FAIL: BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplateStructured/2000pods_100nodes
BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplateStructured/2000pods_200nodes
    scheduler_perf.go:742: disabled by label filter "performance"

=== FAIL: test/integration/scheduler_perf BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplateStructured (0.00s)
--- FAIL: BenchmarkPerfScheduling/SchedulingWithResourceClaimTemplateStructured
    scheduler_perf.go:681: full log output:

=== FAIL: test/integration/scheduler_perf BenchmarkPerfScheduling (0.00s)
--- FAIL: BenchmarkPerfScheduling
```

https://storage.googleapis.com/kubernetes-jenkins/logs/ci-benchmark-scheduler-perf-master/1797657849061445632/build-log.txt

--- a/pkg/scheduler/internal/queue/scheduling_queue.go
+++ b/pkg/scheduler/internal/queue/scheduling_queue.go
@@ -47,6 +47,7 @@ import (
 	"k8s.io/kubernetes/pkg/features"
 	"k8s.io/kubernetes/pkg/scheduler/framework"
 	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/interpodaffinity"
+	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/names"
 	"k8s.io/kubernetes/pkg/scheduler/internal/heap"
 	"k8s.io/kubernetes/pkg/scheduler/metrics"
 	"k8s.io/kubernetes/pkg/scheduler/util"
@@ -1195,9 +1196,10 @@ func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(logger klog.Logger, podIn
 	for _, pInfo := range podInfoList {
 		// Since there may be many gated pods and they will not move from the
 		// unschedulable pool, we skip calling the expensive isPodWorthRequeueing.
-		if pInfo.Gated {
+		if pInfo.Gated && pInfo.UnschedulablePlugins.Has(names.SchedulingGates) {
 			continue
 		}
+
 		schedulingHint := p.isPodWorthRequeuing(logger, pInfo, event, oldObj, newObj)
 		if schedulingHint == queueSkip {
 			// QueueingHintFn determined that this Pod isn't worth putting to activeQ or backoffQ by this event.

--- a/pkg/scheduler/internal/queue/scheduling_queue.go
+++ b/pkg/scheduler/internal/queue/scheduling_queue.go
@@ -1194,8 +1194,14 @@ func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(logger klog.Logger, podIn
 
 	activated := false
 	for _, pInfo := range podInfoList {
-		// Since there may be many gated pods and they will not move from the
-		// unschedulable pool, we skip calling the expensive isPodWorthRequeueing.
+		// Scheduling-gated Pods never get schedulable with any events,
+		// except the Pods themselves got updated, which isn't handled by movePodsToActiveOrBackoffQueue.
+		// So, here we skip them early here so that they don't go through isPodWorthRequeuing,
+		// which isn't fast enough when the number of scheduling-gated Pods in unschedulablePods is large.
+		//
+		// Note that we cannot skip all pInfo.Gated Pods here
+		// because PreEnqueue plugins apart from the scheduling gate plugin may change the gating status
+		// with these events.
 		if pInfo.Gated && pInfo.UnschedulablePlugins.Has(names.SchedulingGates) {
 			continue
 		}

--- a/pkg/scheduler/internal/queue/scheduling_queue.go
+++ b/pkg/scheduler/internal/queue/scheduling_queue.go
@@ -1194,10 +1194,15 @@ func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(logger klog.Logger, podIn
 
 	activated := false
 	for _, pInfo := range podInfoList {
+		// When handling events takes time, a scheduling throughput gets impacted negatively
+		// because of a shared lock within PriorityQueue, which Pop() also requires.
+		//
 		// Scheduling-gated Pods never get schedulable with any events,
 		// except the Pods themselves got updated, which isn't handled by movePodsToActiveOrBackoffQueue.
-		// So, here we skip them early here so that they don't go through isPodWorthRequeuing,
+		// So, we can skip them early here so that they don't go through isPodWorthRequeuing,
 		// which isn't fast enough when the number of scheduling-gated Pods in unschedulablePods is large.
+		// This is a hotfix, which might be changed
+		// once we have a better general solution for the shared lock issue.
 		//
 		// Note that we cannot skip all pInfo.Gated Pods here
 		// because PreEnqueue plugins apart from the scheduling gate plugin may change the gating status

--- a/pkg/scheduler/internal/queue/scheduling_queue.go
+++ b/pkg/scheduler/internal/queue/scheduling_queue.go
@@ -1200,8 +1200,10 @@ func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(logger klog.Logger, podIn
 		// Scheduling-gated Pods never get schedulable with any events,
 		// except the Pods themselves got updated, which isn't handled by movePodsToActiveOrBackoffQueue.
 		// So, we can skip them early here so that they don't go through isPodWorthRequeuing,
-		// which isn't fast enough when the number of scheduling-gated Pods in unschedulablePods is large.
-		// This is a hotfix, which might be changed
+		// which isn't fast enough to keep a sufficient scheduling throughput
+		// when the number of scheduling-gated Pods in unschedulablePods is large.
+		// https://github.com/kubernetes/kubernetes/issues/124384
+		// This is a hotfix for this issue, which might be changed
 		// once we have a better general solution for the shared lock issue.
 		//
 		// Note that we cannot skip all pInfo.Gated Pods here

--- a/pkg/scheduler/internal/queue/scheduling_queue_test.go
+++ b/pkg/scheduler/internal/queue/scheduling_queue_test.go
@@ -1508,11 +1508,9 @@ func TestPriorityQueue_MoveAllToActiveOrBackoffQueueWithQueueingHint(t *testing.
 			expectedQ: unschedulablePods,
 		},
 		{
-			name:    "QueueHintFunction is called when Pod is gated by a plugin other than SchedulingGates",
-			podInfo: setQueuedPodInfoGated(&framework.QueuedPodInfo{PodInfo: mustNewPodInfo(p), UnschedulablePlugins: sets.New("foo")}),
-			hint:    queueHintReturnQueue,
-			// FIXME: This should be backoffQ.
-			// https://github.com/kubernetes/kubernetes/issues/125538
+			name:      "QueueHintFunction is called when Pod is gated by a plugin other than SchedulingGates",
+			podInfo:   setQueuedPodInfoGated(&framework.QueuedPodInfo{PodInfo: mustNewPodInfo(p), UnschedulablePlugins: sets.New("foo")}),
+			hint:      queueHintReturnQueue,
 			expectedQ: activeQ,
 		},
 	}

--- a/pkg/scheduler/internal/queue/scheduling_queue_test.go
+++ b/pkg/scheduler/internal/queue/scheduling_queue_test.go
@@ -44,6 +44,7 @@ import (
 	"k8s.io/kubernetes/pkg/features"
 	"k8s.io/kubernetes/pkg/scheduler/framework"
 	plfeature "k8s.io/kubernetes/pkg/scheduler/framework/plugins/feature"
+	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/names"
 	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/queuesort"
 	"k8s.io/kubernetes/pkg/scheduler/framework/plugins/schedulinggates"
 	"k8s.io/kubernetes/pkg/scheduler/metrics"
@@ -1499,12 +1500,20 @@ func TestPriorityQueue_MoveAllToActiveOrBackoffQueueWithQueueingHint(t *testing.
 			expectedQ: unschedulablePods,
 		},
 		{
-			name:    "QueueHintFunction is not called when Pod is gated",
-			podInfo: setQueuedPodInfoGated(&framework.QueuedPodInfo{PodInfo: mustNewPodInfo(p), UnschedulablePlugins: sets.New("foo")}),
+			name:    "QueueHintFunction is not called when Pod is gated by SchedulingGates plugin",
+			podInfo: setQueuedPodInfoGated(&framework.QueuedPodInfo{PodInfo: mustNewPodInfo(p), UnschedulablePlugins: sets.New(names.SchedulingGates, "foo")}),
 			hint: func(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
 				return framework.Queue, fmt.Errorf("QueueingHintFn should not be called as pod is gated")
 			},
 			expectedQ: unschedulablePods,
+		},
+		{
+			name:    "QueueHintFunction is called when Pod is gated by a plugin other than SchedulingGates",
+			podInfo: setQueuedPodInfoGated(&framework.QueuedPodInfo{PodInfo: mustNewPodInfo(p), UnschedulablePlugins: sets.New("foo")}),
+			hint:    queueHintReturnQueue,
+			// FIXME: This should be backoffQ.
+			// https://github.com/kubernetes/kubernetes/issues/125538
+			expectedQ: activeQ,
 		},
 	}
 
@@ -1518,14 +1527,13 @@ func TestPriorityQueue_MoveAllToActiveOrBackoffQueueWithQueueingHint(t *testing.
 					QueueingHintFn: test.hint,
 				},
 			}
-			test.podInfo.UnschedulablePlugins = sets.New("foo")
 			cl := testingclock.NewFakeClock(now)
 			q := NewTestQueue(ctx, newDefaultQueueSort(), WithQueueingHintMapPerProfile(m), WithClock(cl))
-			// add to unsched pod pool
 			q.activeQ.Add(q.newQueuedPodInfo(test.podInfo.Pod))
 			if p, err := q.Pop(logger); err != nil || p.Pod != test.podInfo.Pod {
 				t.Errorf("Expected: %v after Pop, but got: %v", test.podInfo.Pod.Name, p.Pod.Name)
 			}
+			// add to unsched pod pool
 			err := q.AddUnschedulableIfNotPresent(logger, test.podInfo, q.SchedulingCycle())
 			if err != nil {
 				t.Fatalf("unexpected error from AddUnschedulableIfNotPresent: %v", err)

--- a/test/integration/scheduler/plugins/plugins_test.go
+++ b/test/integration/scheduler/plugins/plugins_test.go
@@ -2668,8 +2668,9 @@ func TestSchedulingGatesPluginEventsToRegister(t *testing.T) {
 	}
 
 	tests := []struct {
-		name              string
-		withEvents        bool
+		name       string
+		withEvents bool
+		// count is the expected number of calls to PreEnqueue().
 		count             int
 		queueHintEnabled  []bool
 		expectedScheduled []bool
@@ -2686,7 +2687,7 @@ func TestSchedulingGatesPluginEventsToRegister(t *testing.T) {
 		{
 			name:              "preEnqueue plugin with event registered",
 			withEvents:        true,
-			count:             2,
+			count:             3,
 			queueHintEnabled:  []bool{false, true},
 			expectedScheduled: []bool{true, true},
 		},
@@ -2700,7 +2701,7 @@ func TestSchedulingGatesPluginEventsToRegister(t *testing.T) {
 			t.Run(tt.name+fmt.Sprintf(" queueHint(%v)", queueHintEnabled), func(t *testing.T) {
 				featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.SchedulerQueueingHints, queueHintEnabled)
 
-				// new plugin every time to clear counts
+				// use new plugin every time to clear counts
 				var plugin framework.PreEnqueuePlugin
 				if tt.withEvents {
 					plugin = &SchedulingGatesPluginWithEvents{SchedulingGates: schedulinggates.SchedulingGates{}}

--- a/test/integration/scheduler/plugins/plugins_test.go
+++ b/test/integration/scheduler/plugins/plugins_test.go
@@ -2651,8 +2651,8 @@ func (pl *SchedulingGatesPluginWOEvents) EventsToRegister() []framework.ClusterE
 	return nil
 }
 
-// This test helps to verify registering nil events for schedulingGates plugin works as expected.
-func TestSchedulingGatesPluginEventsToRegister(t *testing.T) {
+// This test helps to verify registering nil events for PreEnqueue plugin works as expected.
+func TestPreEnqueuePluginEventsToRegister(t *testing.T) {
 	testContext := testutils.InitTestAPIServer(t, "preenqueue-plugin", nil)
 
 	num := func(pl framework.Plugin) int {

